pkg/controllers/management/gke/gke_cluster_handler.go

```diff
@@ -0,0 +1,461 @@
+		DynamicClient:        gkeCCDynamicClient,
+		ClientDialer:         mgmtCtx.Dialer,
+	}}
+
+	wContext.Mgmt.Cluster().OnChange(ctx, "gke-operator-controller", e.onClusterChange)
+}
+
+func (e *gkeOperatorController) onClusterChange(key string, cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil || cluster.Spec.GKEConfig == nil {
+		return cluster, nil
+	}
+
+	if err := e.deployGKEOperator(); err != nil {
+		failedToDeployGKEOperatorErr := "failed to deploy gke-operator: %v"
+		var conditionErr error
+		if cluster.Spec.GKEConfig.Imported {
+			cluster, conditionErr = e.SetFalse(cluster, apimgmtv3.ClusterConditionPending, fmt.Sprintf(failedToDeployGKEOperatorErr, err))
+			if conditionErr != nil {
+				return cluster, conditionErr
+			}
+		} else {
+			cluster, conditionErr = e.SetFalse(cluster, apimgmtv3.ClusterConditionProvisioned, fmt.Sprintf(failedToDeployGKEOperatorErr, err))
+			if conditionErr != nil {
+				return cluster, conditionErr
+			}
+		}
+		return cluster, err
+	}
+
+	// set driver name
+	if cluster.Status.Driver == "" {
+		cluster = cluster.DeepCopy()
+		cluster.Status.Driver = apimgmtv3.ClusterDriverGKE
+		var err error
+		cluster, err = e.ClusterClient.Update(cluster)
+		if err != nil {
+			return cluster, err
+		}
+	}
+
+	// get gke Cluster Config, if it does not exist, create it
+	gkeClusterConfigDynamic, err := e.DynamicClient.Namespace(namespace.GlobalNamespace).Get(context.TODO(), cluster.Name, v1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return cluster, err
+		}
+
+		cluster, err = e.SetUnknown(cluster, apimgmtv3.ClusterConditionWaiting, "Waiting for API to be available")
+		if err != nil {
+			return cluster, err
+		}
+
+		gkeClusterConfigDynamic, err = buildGKECCCreateObject(cluster)
+		if err != nil {
+			return cluster, err
+		}
+
+		gkeClusterConfigDynamic, err = e.DynamicClient.Namespace(namespace.GlobalNamespace).Create(context.TODO(), gkeClusterConfigDynamic, v1.CreateOptions{})
+		if err != nil {
+			return cluster, err
+		}
+
+	}
+
+	gkeClusterConfigMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&cluster.Spec.GKEConfig)
+	if err != nil {
+		return cluster, err
+	}
+
+	// check for changes between gke spec on cluster and the gke spec on the gkeClusterConfig object
+	if !reflect.DeepEqual(gkeClusterConfigMap, gkeClusterConfigDynamic.Object["spec"]) {
+		logrus.Infof("change detected for cluster [%s], updating GKEClusterConfig", cluster.Name)
+		return e.updateGKEClusterConfig(cluster, gkeClusterConfigDynamic, gkeClusterConfigMap)
+	}
+
+	// get gke Cluster Config's phase
+	status, _ := gkeClusterConfigDynamic.Object["status"].(map[string]interface{})
+	phase, _ := status["phase"]
+	failureMessage, _ := status["failureMessage"].(string)
+	if strings.Contains(failureMessage, "403") {
+		failureMessage = fmt.Sprintf("cannot access gke, check cloud credential: %s", failureMessage)
+	}
```

Comment:
Did you test if this works for gke? Like if you add a cloud credential, then delete it in gke, does this catch that?

Reply:
I had not tested it. It does not work, but not because of the message string: https://github.com/rancher/gke-operator/pull/17

Reply:
Okay, once that other PR is merged it is something that should be checked.

Reply:
That other PR is ready to merge, does that mean we can test this now ?

Reply:
Sorry, I should have been more clear. I *had* not tested it but I have now. With the other pull request, it works - there is a string `403` in the error message that can be matched against.